charts/Chart.tsx

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2025 The Ontario Institute for Cancer Research. All rights reserved
+ *
+ * This program and the accompanying materials are made available under the terms of
+ * the GNU Affero General Public License v3.0. You should have received a copy of the
+ * GNU Affero General Public License along with this program.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import { DocumentNode } from 'graphql';
+import { useEffect } from 'react';
+import { useArrangerCharts } from './arranger';
+import { Chart, ChartComponent } from './types';
+
+export type Options = {
+  query: DocumentNode;
+  variables?: Record<string, unknown>;
+  dataTransformer?: (data: unknown) => any;
+};
+
+/**
+ *
+ * Higher order function to return a chart component with a standard end user api and data resolution
+ *
+ * @param generateChartConfig
+ * @param generateChartConfig.Component - A React component for a chart eg. Bar, Line
+ * @param generateChartConfig.options - Options for data resolution
+ * @param generateChartConfig.internalConfig - Additional required config for a chart
+ *
+ * @param chart
+ * @param chart.consumerConfig - A configuration for the UI of a chart
+ * @param chart.onLoad - A callback function for chart loading
+ * @param chart.onError - A callback function for chart errors
+ *
+ * @returns React chart component or null
+ */
+const generateChartComponent =
+  ({
+    // internal
+    Component,
+    options,
+    internalConfig,
+  }: {
+    Component: ChartComponent;
+    options: Options;
+    internalConfig?: Record<string, unknown>;
+  }) =>
```

Comment:
Using inferred generic type for the component so we get type information on the `internalConfig` params.
```suggestion
const generateChartComponent =
  <TComponent extends ChartComponent>({
    // internal
    Component,
    options,
    internalConfig,
  }: {
    Component: TComponent;
    options: Options;
    internalConfig?: Omit<ComponentProps<TComponent>, 'data'>;
  }) =>
```

Reply:
Nice! This is why I added the `ChartComponent` type but React isn't happy with it when I tried to pass a generic. `React.ComponentType` should work but there's some clashes on prop types.